### PR TITLE
[RFC] module versioning

### PR DESCRIPTION
**Pull request by psteckler**
Proposal for extending the module versioning discipline for modules that have an externally-representable type.

### CODE REVIEW

**Comment by bkase** [comment]
Also can you mention avoiding non stable models from being serializavle (I think sexp is okay, but no yojson nor binio)

File: rfcs/0013-module-versioning.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+A use of the ppx might look like:
+
+```ocaml
+[@@versioned_module
+  (`Version 1
+  ,`Type int * string,
+  ,`Serializations [bin_io;sexp;yojson]
+  )
+]
+```
```
Comment (bkase) [inline rfcs/0013-module-versioning.md]: How will this support other deriving bits? `eq` and `compare` for example.

Comment (bkase) [reply rfcs/0013-module-versioning.md]: but I like this idea to reduce the boilerplate

**Comment by enolan** [comment]
Are you intentionally leaving how to migrate old versions of data into the new representations out of scope? That seems important for the situations when we're deserializing data from old versions of Coda.

File: rfcs/0013-module-versioning.md (hunk 1 above)
Comment (psteckler) [reply rfcs/0013-module-versioning.md]: I think we can add as many as we like, but only the serializations will trigger additional code.

**Comment by psteckler** [comment]
> Are you intentionally leaving how to migrate old versions of data into the new representations out of scope? 

Yes. A writer of a new version can add such a transformation, if needed.

**Comment by psteckler** [comment]
Added text in response to comments by @enolan and @bkase.

File: rfcs/0013-module-versioning.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+        (* similarly for binio, sexp *)
+
+      end
+    end
+  end
+end
+```
+
+Putting the name and version in the type `Versioned.t` means they'll be
```
Comment (imeckler) [inline rfcs/0013-module-versioning.md]: Doesn't the versioned RPC create its own "versioned" objects? What is the need to define them explicitly out of the RPC library? For converting on-disk values perhaps?

File: rfcs/0013-module-versioning.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+It could be useful to transform data from an older version to a newer
+version. Such a transformer could be included as a ppx argument.
+
+### Serialization restricted to Stable, versioned modules
+
+Serialization should be allowed only for modules following
+this discipline, including the use of the `Stable` submodule
+containing versions.
+
+An exception can be carved out for using `sexp` serialization of data
```
Comment (imeckler) [inline rfcs/0013-module-versioning.md]: 👍

File: rfcs/0013-module-versioning.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+]
+```
+The ppx would generate the `Version` module, as above. That module would not
+contain any special code for `eq` or other non-serialization `deriving` items,
+other than to mention them in a `deriving` annotation.
+
+## Drawbacks
+
+We may not need the versioning for `sexp` and `yojson`, if these representations
+are not, in fact, used by other software.
```
Comment (imeckler) [inline rfcs/0013-module-versioning.md]: I think we probably only need the versioning for `bin_io`? The two main use cases I can think of are RPC communication with nodes running different versions of the software and being able to run a new version of the software with on-disk state created by an older version, and migrating data over to the new format.

File: rfcs/0013-module-versioning.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+producing and consuming representations for older versions of the software.
+
+Suppose we create a module with a name and a version:
+
+```ocaml
+module Some_data = struct
+
+  let name = "some_data"
+
+  module Stable = struct
```
Comment (imeckler) [inline rfcs/0013-module-versioning.md]: Have you taken a look at `Async.Versioned_rpc` and made sure we do something that's easy to use with that?

File: rfcs/0013-module-versioning.md (hunk 2 above)
Comment (psteckler) [reply rfcs/0013-module-versioning.md]: Say we returned a JSON representation of a versioned type via a REST API.

I don't know what the exact use cases might be, tho. In the `Drawbacks` section, I mention the uncertainy of the need for this.

File: rfcs/0013-module-versioning.md (hunk 4 above)
Comment (psteckler) [reply rfcs/0013-module-versioning.md]: The versioning of modules here is independent of the RPC versioning, which is the subject of RFC 0012.

If you have a `bin_io` representation of data, then you can turn it into OCaml data, and do the migration between versions in OCaml.

File: rfcs/0013-module-versioning.md (hunk 5 above)
Comment (psteckler) [reply rfcs/0013-module-versioning.md]: It looks like the query and response types need to have `bin_io` there, which is compatible with the regime here.

But again, I'm thinking of this module versioning as distinct from the RPC versioning, which has its own mechanisms.

File: rfcs/0013-module-versioning.md (hunk 4 above)
Comment (psteckler) [reply rfcs/0013-module-versioning.md]: I spoke with @bkase about this, and realized a type in a versioned module could be mentioned in an RPC type. I added a para about that.

File: rfcs/0013-module-versioning.md (hunk 5 above)
Comment (psteckler) [reply rfcs/0013-module-versioning.md]: See my response to the earlier comment about RPC.

**Pull request merged**